ServerLayer/Source/ServerLayer.cpp

```diff
@@ -126,151 +123,193 @@ void ServerLayer::receiveFromClients()
+void ServerLayer::sendData(const std::string& data, const Whereto recipient) const
 {
-    _isRunning.store(true);
-    _isRunningForClient.store(true);
-    waitLoop();
+    switch (recipient)
+    {
+        case Whereto::SERVER:
+            emit signalToSendToServer(data.c_str());
+            break;
+        
+        case Whereto::CLIENT:
+            emit signalToSendToClient(data.c_str());
+            break;
+        
+        default:
+            assert(false);
+            break;
+    }
 }
 
 void ServerLayer::launch()
 {
-    initSocket(_layerSocket);
-    bindSocket(_layerSocket, _layerSocketAddress, _layerPort);
-
-    listenOn(_layerSocket, _backlog);
+    if (_layerSocket->listen(QHostAddress::Any, _layerPort))
+    {
+        _printer.writeLine(std::cout, "Layer server is started!");
+    }
+    else
+    {
+        _printer.writeLine(std::cout, "Layer server is not started!");
+    }
     doConnection();
 }
 
-void ServerLayer::waitingForConnections()
+bool ServerLayer::tryConnect(const int port, const std::string& ip, QTcpSocket* socketToConnect,
+                             const int msecs) const
 {
-    closeSocket(_clientSocket);
-    _isRunningForClient.store(false);
-    process();
-}
+    if (socketToConnect->isOpen())
+    {
+        socketToConnect->close();
+    }
 
-std::string ServerLayer::getServerIP() const noexcept
-{
-    return _serverIP;
-}
+    socketToConnect->connectToHost(ip.c_str(), port);
 
-void ServerLayer::setServerIP(const std::string_view newServerIP)
-{
-    _serverIP = newServerIP;
+    if (socketToConnect->waitForConnected(msecs))
+    {
+        _printer.writeLine(std::cout, "Connected to Server!");
+        return true;
+    }
+
+    _printer.writeLine(std::cout, "Not connected to Server!");
+    return false;
 }
 
-void ServerLayer::tryReconnectToServer()
+void ServerLayer::doConnection()
 {
-    while (!_isRunning.load())
+    // Remember that sending socket should be connect to server receiving port
+    // and receiving socket should be connect to server sending port!
+    bool isConnected = tryConnect(_serverReceivingPort, _serverIP, _sendingSocket.get())
+                    && tryConnect(_serverSendingPort, _serverIP, _receivingSocket.get());
+    while (!isConnected)
     {
-        closeSocket(_sendingSocket);
-        closeSocket(_receivingSocket);
-
-        initSocket(_sendingSocket);
-        initSocket(_receivingSocket);
-
-        doConnection();
+        isConnected = tryConnect(_serverReceivingPort, _serverIP, _sendingSocket.get())
+                   && tryConnect(_serverSendingPort, _serverIP, _receivingSocket.get());
 
         std::this_thread::sleep_for(
             std::chrono::milliseconds(CONFIG.get<CAST(Param::RECONNECTION_DELAY)>()));
     }
+
+    _printer.writeLine(std::cout, "\nServerLayer launched...\n");
+    _logger.writeLine("\nServerLayer launched at", utils::getCurrentSystemTime());
+}
+
+void ServerLayer::tryReconnectToServer()
+{
+    doConnection();
 }
```

Comment:
Do you need this metod.

Reply:
Yes, because not every reconnection is equal to connection. And I tried to highlight the main methods for the future network interface. And what method do I need to call if I cannot know realization of the ServerLayer? Yes, the absence of such a method is misleading.